### PR TITLE
Support update artifact/execution type in metadata

### PR DESCRIPTION
**Pull request by dreamryx**
**What this PR does / why we need it**:
Supporting update artifact/execution type in metadata. For now, the update only support add fields.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Release note**:

```release-note

```

---
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/metadata/110)

### CODE REVIEW

File: api/service.proto [hunk 1]
```diff
@@ -35,6 +35,24 @@ message CreateArtifactTypeResponse {
   ml_metadata.ArtifactType artifact_type = 1;
 }
 
+message PutTypeOptions {
```
Comment (zhenghuiwang) [inline api/service.proto]: Can we rename PutTypeOptions to `UpdateTypeOptions`? `Put` is too specific to HTTP verb.

But I don't think we need it. See my comment below.

File: service/service.go [hunk 2]
```diff
@@ -158,6 +158,51 @@ func (s *Service) CreateArtifactType(ctx context.Context, req *api.CreateArtifac
...
+
+	if req.ArtifactType == nil {
+		return nil, errors.New("no ArtifactType specified")
+	}
+
+	if req.PutTypeOptions == nil {
+		return nil, errors.New("no PutTypeOptions specified")
+	}
+
+	if req.PutTypeOptions.CanDeleteFields {
```
Comment (zhenghuiwang) [inline service/service.go]: The only acceptable PutTypeOptions now is

```
putTypeOptions {
					CanAddFields:       true
					CanDeleteFields:    false
					AllFieldsMustMatch: true}
```

How about not exposing PutTypeOptions in the API and add it back later if more possible configs are supported?

File: service/service_test.go [hunk 3]
```diff
@@ -203,6 +203,95 @@ func TestCreateArtifactType(t *testing.T) {
 	}
 }
 
+func TestUpdateArtifactType(t *testing.T) {
```
Comment (zhenghuiwang) [inline service/service_test.go]: Can you also write this unit test as integration test? The integration script is at

https://github.com/kubeflow/metadata/blob/master/test/e2e/make_requests.sh

The presubmit  job will start a server and make these requests.

File: service/service.go (hunk 2 above)
Comment (dreamryx) [reply service/service.go]: I'm OK with it. We shall expose this later.

File: service/service_test.go (hunk 3 above)
Comment (dreamryx) [reply service/service_test.go]: Yes. Of course. Thanks for the comment.

File: api/service.proto (hunk 1 above)
Comment (dreamryx) [reply api/service.proto]: We shall expose it later. But the update function does not expose the option looks a bit weird for now.

File: service/service.go [hunk 4]
```diff
@@ -158,6 +158,34 @@ func (s *Service) CreateArtifactType(ctx context.Context, req *api.CreateArtifac
...
+	var putTypeOptions mlmetadata.PutTypeOptions
+
+	if req.ArtifactType == nil {
+		return nil, errors.New("no ArtifactType specified")
+	}
+
+	putTypeOptions.AllFieldsMustMatch = true
+	putTypeOptions.CanAddFields = true
+
+	_, err := s.store.PutArtifactType(req.ArtifactType, &putTypeOptions)
```
Comment (zhenghuiwang) [inline service/service.go]: It is more readable if you can use struct literal:
```
s.store.PutArtifactType(req.ArtifactType, &putTypeOptions {
   AllFieldsMustMatch : true,
   CanAddFields: true,
})
```

File: service/service.go [hunk 5]
```diff
@@ -310,6 +338,34 @@ func (s *Service) CreateExecutionType(ctx context.Context, req *api.CreateExecut
...
+	var putTypeOptions mlmetadata.PutTypeOptions
+
+	if req.ExecutionType == nil {
+		return nil, errors.New("no ExecutionType specified")
+	}
+
+	putTypeOptions.AllFieldsMustMatch = true
+	putTypeOptions.CanAddFields = true
+
+	_, err := s.store.PutExecutionType(req.ExecutionType, &putTypeOptions)
```
Comment (zhenghuiwang) [inline service/service.go]: Please use struct literal here.

Comment (dreamryx) [reply service/service.go]: OK.

File: service/service.go (hunk 4 above)
Comment (dreamryx) [reply service/service.go]: OK. No problem.

**Comment by zhenghuiwang** [comment]
/lgtm

@neuromage any comments?

**Comment by dreamryx** [comment]
@zhenghuiwang Thanks for your review! :)

**Pull request merged**